python/cudf/cudf/core/udf/groupby_typing.py

```diff
@@ -153,15 +176,86 @@ def caller(data, index, size):
+        # early exit to make sure typing doesn't fail for normal
+        # non-group ops
+        if not all(isinstance(arg, GroupType) for arg in args):
+            return None
+        # check if any groups are poisoned for this op
+        for arg in args:
+            if isinstance(arg.group_scalar_type, types.Poison):
+                raise UDFError(
+                    f"Use of a column of {arg.group_scalar_type.ty} detected "
+                    "within UDF body. \nOnly columns of the following dtypes "
+                    "may be used through the \nGroupBy.apply() JIT engine: "
+                    f"{[str(x) for x in SUPPORTED_GROUPBY_NUMPY_TYPES]}"
+                )
+        fname = self.key.__name__
+        if fname in call_cuda_functions:
+            for sig in call_cuda_functions[fname].keys():
+                if all(
+                    arg.group_scalar_type == ty for arg, ty in zip(args, sig)
+                ):
+                    return nb_signature(sig[0], *args)
+        raise UDFError(self.make_error_string(args))
+
+
+class GroupAttrBase(AbstractTemplate):
+    def make_error_string(self, args):
+        fname = self.key.split(".")[-1]
+        args = (self.this, *args)
+        dtype_err = ", ".join([str(g.group_scalar_type) for g in args])
+        sr_err = ", ".join(["Series" for _ in range(len(args) - 1)])
+        return (
+            f"\nSeries.{fname}({sr_err}) is not supported for "
+            f"({dtype_err}) within JIT GroupBy apply.\nTo see "
+            f"what's available, visit {_UDF_DOC_URL}"
+        )
 
-        def generic(self, args, kws):
-            for retty, inputty in call_cuda_functions[funcname.lower()].keys():
-                if self.this.group_scalar_type == inputty:
-                    return nb_signature(retty, recvr=self.this)
+    def generic(self, args, kws):
+        # earlystop to make sure typing doesn't fail for normal
+        # non-group ops
+        if not all(isinstance(arg, GroupType) for arg in args):
             return None
+        # check if any groups are poisioned for this op
+        for arg in (self.this, *args):
+            if isinstance(arg.group_scalar_type, types.Poison):
+                raise UDFError(
+                    f"Use of a column of {arg.group_scalar_type.ty} detected "
+                    "within UDAF body. \nOnly columns of the following dtypes "
+                    "may be used through the \nGroupBy.apply() JIT engine: "
+                    f"{[str(x) for x in SUPPORTED_GROUPBY_NUMPY_TYPES]}"
+                )
+        fname = self.key.split(".")[-1]
+        for sig in call_cuda_functions[fname].keys():
+            retty, selfty, *argtys = sig
+            if self.this.group_scalar_type == selfty and all(
+                arg.group_scalar_type == ty for arg, ty in zip(args, argtys)
+            ):
+                return nb_signature(retty, *args, recvr=self.this)
+        raise UDFError(self.make_error_string(args))
```

Comment:
Same suggestions from above apply here too for iterating over args and using the walrus.

Reply:
Done